### PR TITLE
introduce the ability to mock time.

### PR DESCRIPTION
**Pull request by raulk**
This PR introduces the ability to mock time, via a fork of https://github.com/benbjohnson/clock that I created under https://github.com/raulk/clock (can migrate to filecoin-project org if needed). The fork was needed to introduce the `Until` method.

This is important for Testground tests, where we want to control and coordinate miners by synchronizing their mining loops so we can create all sorts of convoluted chain shapes and forks. We have tried reducing the block delay and propagation delay times, but we need a more deterministic method.

Keeping the mainnet parameters untouched (block and propagation delay, 25s and 6s respectively), but instead controlling the clock and synchronizing all instances via the Testground sync service, seems like the way to go!

I have left tests and CLI tooling untouched, as those probably have no need to ever use mocked time.

### CODE REVIEW

**Comment by raulk** [comment]
I've integrated this downstream in Project Oni, and it's working pretty well: https://github.com/filecoin-project/oni/pull/143. Flagging this as ready for review, @magik6k.

File: chain/blocksync/blocksync.go [hunk 1]
```diff
@@ -126,7 +127,7 @@ func (bss *BlockSyncService) HandleStream(s inet.Stream) {
 	}
 
 	writeDeadline := 60 * time.Second
-	_ = s.SetDeadline(time.Now().Add(writeDeadline))
+	_ = s.SetDeadline(build.Clock.Now().Add(writeDeadline))
```
Comment (Kubuxu) [inline chain/blocksync/blocksync.go:130]: Socket/stream deadlines should use real time as internally they use real time.

File: chain/blocksync/blocksync_client.go [hunk 2]
```diff
@@ -284,25 +285,25 @@ func (bs *BlockSync) fetchBlocksBlockSync(ctx context.Context, p peer.ID, req *B
...
 
-	start := time.Now()
+	start := build.Clock.Now()
 	s, err := bs.host.NewStream(inet.WithNoDial(ctx, "should already have connection"), p, BlockSyncProtocolID)
 	if err != nil {
 		bs.RemovePeer(p)
 		return nil, xerrors.Errorf("failed to open stream to peer: %w", err)
 	}
-	_ = s.SetWriteDeadline(time.Now().Add(5 * time.Second))
+	_ = s.SetWriteDeadline(build.Clock.Now().Add(5 * time.Second))
```
Comment (Kubuxu) [inline chain/blocksync/blocksync_client.go:294]: Same with deadline here.

File: miner/miner.go [hunk 3]
```diff
@@ -170,14 +170,18 @@ func (m *Miner) mine(ctx context.Context) {
...
-					time.Sleep(time.Until(btime))
+					build.Clock.Sleep(build.Clock.Until(btime))
 				}
-			} else {
-				log.Warnw("mined block in the past", "block-time", btime,
-					"time", time.Now(), "duration", time.Since(btime))
+			default:
+				log.Warnw("mined block in the past",
+					"block-time", btime, "time", build.Clock.Now(), "difference", build.Clock.Since(btime))
 			}
```
Comment (raulk) [inline miner/miner.go:185]: I slid in a log refinement here. With mocked time, we deterministically hit the edge case of block time == clock time, which could also happen with real time (albeit very sporadically). The "mined block in the past" message is inaccurate in those situations.

File: chain/blocksync/blocksync_client.go (hunk 2 above)
Comment (raulk) [reply chain/blocksync/blocksync_client.go]: Fixed in [`fbf6596` (#2362)](https://github.com/filecoin-project/lotus/pull/2362/commits/fbf6596f20744628212d4b6a5c92ded32ae0693f).

File: chain/blocksync/blocksync.go (hunk 1 above)
Comment (raulk) [reply chain/blocksync/blocksync.go]: Fixed in [`fbf6596` (#2362)](https://github.com/filecoin-project/lotus/pull/2362/commits/fbf6596f20744628212d4b6a5c92ded32ae0693f).

**Comment by raulk** [comment]
@magik6k @Kubuxu ping

File: miner/miner.go (hunk 3 above)
Comment (schomatis) [reply miner/miner.go:185]: Seems fine.

**Pull request merged**